Collapsible records in playground output

berekuk opened the pull request:
Fixes #767.

This is a draft, just sharing the work so far. The main issue is that I can't get animations right - [Framer Motion](https://www.framer.com/motion/) messes up transforms and I can't figure out why. Various combinations of `layout`, `layout="position"` and `<LayoutGroup>` cause various kinds of issues.

 I guess I should give up and remove all animations for now, since I've spent 3 hours fighting this already, that's too much.

berekuk commented:
Example: https://deploy-preview-796--squiggle-documentation.netlify.app/playground#code=eNpdjM8KgkAQxl9lmVPGXhK8LHSKnsCOe1lp1YVxptaVNPXdG6wONXww8Pv%2BzNC3%2FLhMN9%2BDSXHwegMnphQZhdUO%2Bw8sh65zcfph52tIHL%2FVQCEFh%2BV9CE2DvkwxUAMGRnVUhaVJ3mxJydXMRs3LuIz7YrUkeopJHDuHu1wXmaX0l7bQekS2oN%2BwctGoyie3O%2Bg82zYkJ6XcEqwvm39I2w%3D%3D

Note that the issue with animations is not just the awkward position _during_ the animation (that's fixable) but that _after_ the animation scale can become broken.

Requests for feedback:
- does the new style look good or does it seem too minimalistic?
- do we even need `showTypes` setting? I think we could just always display type on hover and that would be ok
- aligning everything to the left seems better than before, but for arrays it's a bit hard to distinguish keys and values if values are numbers too

Further ideas:
- highlight the ancestors tree (names and left borders) on hover?
- maybe draw the left border for single values too? that'd help with the arrays issue above
- cache the collapse state somehow so that it'd be remembered on code edits (right now collapse state is remembered if the new version of code compiles, but resets on compilation errors)

OAGr commented:
Looks promising!

Some quick thoughts:
1. First version, I don't think we need animations, especially if they seem like a pain. We'll have lots of time to iterate.
2. We could probable remove ``showTypes``. It was really made for a previous time, I don't think any of us use it now. The implementation isn't well done, for one. 
3. I have some ideas about UI; I'd suggest focusing on functionality from here, then I'll take a go at the UI later.

berekuk commented:
Lots of changes!

- local settings for charts
- modal windows
- reused ViewSettings for local settings which required some further refactorings
- small win: removed showControls setting

This got a bit out of hand in scope and complexity, but I'd like to argue in favor of keeping all this.

On "plot options will control chart settings, so we shouldn't add too many non-persistent settings" argument: that's true and keeping settings is code is better, but we have already have view settings in the playground. And it doesn't make sense to keep most of them global. Also, I think it's quite probable that frontend part will be ahead of language features for some time, and there might be some settings for which configuring through code is too unhandy (e.g. zooming and panning).

On merging the settings. There's the UI insight which I didn't realize at first: nested settings are hard to use because you might want to change your global settings after you changed your local settings.

E.g., if there's a tree:
- top-level
- `foo` record
- `foo.bar` array
- `foo.bar[0]`, `foo.bar[1]`, `foo.bar[2]` charts

It might make sense to allow to configure _all_ `foo.bar` charts at `foo.bar` level. But then if you go to the top-level playground settings and change something, it won't affect your charts, because now they're controlled at the `foo.bar` level. And if the user also changed something at `foo` level and at `foo.bar[2]` level then it would be difficult for them to figure out why their settings don't apply.

So, I had to drop the idea of making settings truly nested, and opted for leaf-level settings for charts instead.

This still requires the "Reset settings" button, but I hope it's obvious enough:
<img width="625" alt="Screenshot 2022-07-22 at 23 45 06" src="https://user-images.githubusercontent.com/89368/180516326-97564b8b-1a07-46c5-af61-57017653f2c4.png">

berekuk commented on packages/components/src/components/SquiggleViewer/ItemSettingsMenu.tsx:
```diff
@@ -0,0 +1,129 @@
+      distributionChartActions: mergedSettings.distributionPlotSettings.actions,
+      diagramStart: mergedSettings.chartSettings.start,
+      diagramStop: mergedSettings.chartSettings.stop,
+      diagramCount: mergedSettings.chartSettings.count,
```
Note: I really hate this, but it's hard to clean up without significant refactorings. All props in Playground, Chart, etc. should be refactored and unified, then we could pass around objects with common shapes instead of uniquely shaped objects with sometimes-different keys.

OAGr commented:
I Slack sent you some recordings about this, where I go more into detail on it.

OAGr reviewed (changes requested):
I think the UI should get some cleanup before this goes in. I sent a separate video which describes that.

berekuk commented:
Changes:
* escape key for modals
* close modal by clicking on the background
* bar color is blue now
* smart modal positioning (maybe even too smart... made many attempts to get this right, lots of corner cases & CSS/DOM quirks)
* fixed tooltip styles
* small semi-hidden feature: when modal is open, you can scroll back to the item you're configuring by clicking on item in modal's title; but the link is not highlighted (since I'm unsure if that would maybe be too annoying for a pretty rare scenario)

On modal positioning logic:
* modal tries to position itself in the middle of the _visible_ part of the left sidebar
* unless the rectangle where it tries to position itself is too small; then it falls back to positioning in the middle of the screen

I decided not to remove the gray overlays (but made them lighter); making local settings non-modal, i.e. making stuff outside of settings forms interactive, would cause too many UI scenarios I'm not sure how to handle, or which cause too many complications for this to be worth it.

Also, I kept the overlays _global_. I.e., if playground is embedded into a larger website, overlays cover the entire screen and not just the part that's controlled by the playground. The reason for this is that modal's position should be fixed, while the larger website could be scrollable, and keeping the modal inside the playground boundaries would cause problems when the playground gets almost-out-of-viewport on scroll. (This could be handled with switching from `position: fixed` to `position: relative` dynamically, so that modal would "give up" on being fixed entirely on screen and leave the screen; but it's complicated).

Btw, double tooltip border was not due to tailwind, it's because docusaurus redefines `<pre>` styles. I got rid of `<pre>` and replaced it with equivalent css styles.

Remaining issues:
* height is still broken (but that's not a regression, it's broken on develop branch too)
* top menu on the website covers the modal & tooltips; I blame docusaurus, its navbar is fixed with huge z-index, so there's not much I can do; I've enabled [hideOnScroll](https://docusaurus.io/docs/api/themes/configuration#auto-hide-sticky-navbar) feature to fix this, but overlaps still can happen occasionally
  * on preview branches netlify also adds a bottom bar, but that won't be a problem after the merge

OAGr commented:
Small note: The editor shows "Distribution Display Settings" and "Function Display Settings" for each item. It would be nice if it just showed distribution settings for distributions, and function settings for functions. Maybe also show "Distribution Display Settings" for functions, for the scroll, but have them on bottom.

OAGr commented:
I think the model form could be 1.5x-2x the width. 
<img width="747" alt="image" src="https://user-images.githubusercontent.com/377065/181087092-fbf617b0-ad07-4c6f-b5f0-771817098ac3.png">

OAGr commented:
I think this is a fair bit better. I'm inclined to try to get it into develop, then production, soon; we have several areas of improvement to do, before we get into more serious polish .

berekuk commented:
> I think the model form could be 1.5x-2x the width.

Ok, increased.

I think this is ready to be merged; I'll fix the height bug with separate PR, created #886 for that.

OAGr commented:
One small thing still: When I click the keys, they still get highlighted in blue. Did you look into this? (We can adjust it later) 
<img width="182" alt="image" src="https://user-images.githubusercontent.com/377065/181132253-6d39f6cd-2bbc-402e-a600-aaefe74e456d.png">

OAGr commented:
We could really use this functionality soon (to help with speed), and it generally works pretty well, imo. I didn't do a thorough code review, but trust you enough here. (And feel like we might replace/remove some of this later anyway, not sure).
